File: heritage.cpp
/*
ID: totheso1
LANG: C++14
TASK: heritage
*/

#include <algorithm>
#include <fstream>
#include <iterator>
#include <string>

using namespace std;

static ifstream fin("heritage.in");
static ofstream fout("heritage.out");

#define endl '\n'

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(const int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

static const auto in_order = fin_get<string>();
static const auto pre_order = fin_get<string>();

template <typename InputIt>
static void print_post_order_recursive(const InputIt in_l, const InputIt in_r,
                                       const InputIt pre_l,
                                       const InputIt pre_r) {
  if (in_l == in_r && pre_l == pre_r) {
    return;
  }
  const auto root_in = find(in_l, in_r, *pre_l);
  const auto right_pre = next(pre_l, 1 + distance(in_l, root_in));
  print_post_order_recursive(in_l, root_in, next(pre_l), right_pre);
  print_post_order_recursive(next(root_in), in_r, right_pre, pre_r);
  fout << *pre_l;
}

int main() {
  print_post_order_recursive(cbegin(in_order), cend(in_order),
                             cbegin(pre_order), cend(pre_order));
  fout << endl;
}

File: heritage.py
"""
ID: totheso1
LANG: PYTHON3
TASK: heritage
"""

from __future__ import annotations

from dataclasses import dataclass
from itertools import islice
from typing import Any, Optional

fin = open("heritage.in", "r")
fout = open("heritage.out", "w")


def fprint(*args: Any, **kwargs: Any) -> None:
    print(*args, file=fout, **kwargs)


# Code start
in_order = fin.readline().strip()
pre_order = fin.readline().strip()


def solve() -> None:
    def print_post_order_recursive(
        in_l: int, in_r: int, pre_l: int, pre_r: int
    ) -> None:
        if in_l == in_r and pre_l == pre_r:
            return
        root_in = in_order.find(pre_order[pre_l], in_l, in_r)
        right_pre = pre_l + root_in - in_l + 1
        print_post_order_recursive(in_l, root_in, pre_l + 1, right_pre)
        print_post_order_recursive(root_in + 1, in_r, right_pre, pre_r)
        fprint(pre_order[pre_l], end="")

    print_post_order_recursive(0, len(in_order), 0, len(pre_order))
    fprint()


solve()
# Code end
